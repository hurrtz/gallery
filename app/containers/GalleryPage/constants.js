export const FETCH_CAR = 'FETCH_CAR';
export const SET_CAR = 'SET_CAR';

export const FETCH_CAR_URL = 'https://www.mobile.de/hiring-challenge.json';
export const CAR_IMAGE_FILE_EXTENSION = 'jpg';
