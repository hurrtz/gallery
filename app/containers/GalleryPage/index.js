import React from 'react';

export default function GalleryPage() {
  return <h1>GalleryPage</h1>;
}
