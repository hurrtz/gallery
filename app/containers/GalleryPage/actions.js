import { FETCH_CAR, SET_CAR } from './constants';

export const fetchCar = () => ({ type: FETCH_CAR });
export const setCar = car => ({ type: SET_CAR, payload: car });
