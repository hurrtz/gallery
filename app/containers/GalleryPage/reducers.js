import { SET_CAR } from './constants';

export const initialState = {
  car: {
    id: undefined,
    title: '',
    imageURLs: [],
  },
};

const galleryReducer = (state = initialState, action) => {
  switch (action.type) {
    case SET_CAR:
      return { ...state, car: action.payload };

    default:
      return state;
  }
};

export default galleryReducer;
