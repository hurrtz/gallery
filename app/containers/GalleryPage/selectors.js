import { createSelector } from 'reselect';
import { initialState } from './reducers';

const selectState = state => state.gallery || initialState;

const makeSelectCar = () =>
  createSelector(
    selectState,
    state => state.car || undefined,
  );

const makeSelectCarTitle = () =>
  createSelector(
    [makeSelectCar()],
    car => (car ? car.title : ''),
  );

const makeSelectCarImageURLs = () =>
  createSelector(
    [makeSelectCar()],
    car => (car ? car.imageURLs : []),
  );

export { makeSelectCar, makeSelectCarTitle, makeSelectCarImageURLs };
