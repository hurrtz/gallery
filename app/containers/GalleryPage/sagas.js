import { put, select, takeLatest } from 'redux-saga/effects';

import { FETCH_CAR, FETCH_CAR_URL } from './constants';
import { setCar } from './actions';
import { makeSelectCar } from './selectors';

export function* fetchCar() {
  const car = yield select(makeSelectCar());

  if (car) {
    try {
      // map the response to something the reducer can work with
      const fetchedCar = yield fetch(FETCH_CAR_URL)
        .then(response => response.json())
        .then(response => ({
          id: response.id,
          title: response.title,
          imageURLs: [...new Set(response.images.map(image => image.uri))],
        }));

      if (fetchedCar) {
        yield put(setCar(fetchedCar));
      }
    } catch (error) {
      console.error(error);
    }
  }
}

export default function* gallerySagas() {
  yield takeLatest(FETCH_CAR, fetchCar);
}
