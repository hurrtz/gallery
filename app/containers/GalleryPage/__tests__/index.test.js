import React from 'react';
import { Provider } from 'react-redux';
import { ConnectedRouter } from 'connected-react-router';
import history from 'utils/history';
import { render } from 'react-testing-library';

import GalleryPage from '../index';
import configureStore from '../../../configureStore';

const initialState = {};
const store = configureStore(initialState, history);

describe('<GalleryPage />', () => {
  it('should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(
      <Provider store={store}>
        <ConnectedRouter history={history}>
          <GalleryPage />
        </ConnectedRouter>
      </Provider>,
    );
    expect(firstChild).toMatchSnapshot();
  });
});
