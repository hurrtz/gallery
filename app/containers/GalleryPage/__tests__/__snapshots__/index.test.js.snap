// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GalleryPage /> should render and match the snapshot 1`] = `
<div
  class="MuiContainer-root makeStyles-section-1 MuiContainer-maxWidthLg"
>
  <h2
    class="MuiTypography-root makeStyles-card-4 MuiTypography-h2"
  >
    <span>
      â€”
    </span>
  </h2>
  <h3
    class="MuiTypography-root makeStyles-card-4 MuiTypography-h3"
  >
    Picture Gallery
  </h3>
  <hr
    class="MuiDivider-root makeStyles-divider-3"
  />
  <div
    class="MuiContainer-root makeStyles-pictureGalleryWrapper-2 makeStyles-section-1 MuiContainer-maxWidthLg"
  />
</div>
`;
